### PR TITLE
Batch

### PR DESCRIPTION
**Pull request by theodorewahle**

### CODE REVIEW

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  static GsonBuilder gson = new GsonBuilder().setFieldNamingStrategy((field) -> toCamelCase(field.getName()));
+
+
+  public <T extends ApiData> SmartcarResponse<T> get(String path) throws SmartcarException {
+    JsonObject res = this.data.get(path);
+    Integer code = res.get("code").getAsInt();
+    if (code != 200) {
+      // TODO : handle error
+      System.out.println("ERROR");
```
Comment (theodorewahle) [inline src/main/java/com/smartcar/sdk/data/BatchResponse.java]: @sankethkatta do we want to throw one type of error here or many different kinds?

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      case "/odometer":
+        return (Class<T>) VehicleOdometer.class;
+      case "/battery":
+        return (Class<T>) VehicleBattery.class;
+      case "/charge":
+        return (Class<T>) VehicleCharge.class;
+      default:
+        return (Class<T>) VehicleFuel.class;
+      }
+  }
```
Comment (theodorewahle) [inline src/main/java/com/smartcar/sdk/data/BatchResponse.java]: @sankethkatta do you think there is a better way to do this than a straight-up switch statement? Should I throw if they ask for a path that is invalid?

File: build.gradle [hunk 3]
```diff
@@ -2,7 +2,7 @@ plugins {
     id 'com.adarshr.test-logger' version '1.3.1'
     id 'com.avast.gradle.docker-compose' version '0.8.0'
     id 'com.jfrog.bintray' version '1.8.0'
-    id 'org.unbroken-dome.test-sets' version '1.5.0'
+    id 'org.unbroken-dome.test-sets' version '2.0.0'
```
Comment (theodorewahle) [inline build.gradle]:
```suggestion
    id 'org.unbroken-dome.test-sets' version '1.5.0'
```

File: src/integration/java/com/smartcar/sdk/VehicleTest.java [hunk 4]
```diff
@@ -137,6 +139,23 @@ public void testActionUnlock() throws SmartcarException {
         this.vehicle.unlock();
     }
 
+    /**
+     * Tests that the batch request method works.
+     */
+    @Test(groups = "vehicle")
+    public void testBatch() throws SmartcarException {
+        ArrayList<String> paths = new ArrayList<>();
```
Comment (sankethkatta) [inline src/integration/java/com/smartcar/sdk/VehicleTest.java]: Normally, in Java you set the type of the variable to be the interface class.

```suggestion
        List<String> paths = new ArrayList<>();
```

File: src/main/java/com/smartcar/sdk/Vehicle.java [hunk 5]
```diff
@@ -7,24 +7,29 @@
 
 import javax.json.Json;
 import javax.json.JsonObject;
+import javax.sound.midi.SysexMessage;
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/Vehicle.java]: This import seems to be unused.

File: src/main/java/com/smartcar/sdk/Vehicle.java [hunk 6]
```diff
@@ -7,24 +7,29 @@
...
+import javax.json.JsonArray;
+import javax.json.JsonArrayBuilder;
 
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Arrays;
 
+import okio.Buffer;
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/Vehicle.java]: This import seems unused.

File: src/main/java/com/smartcar/sdk/Vehicle.java [hunk 7]
```diff
@@ -7,24 +7,29 @@
...
+import java.io.IOException;
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/Vehicle.java]: This also seems unused.

File: src/main/java/com/smartcar/sdk/Vehicle.java [hunk 8]
```diff
@@ -288,6 +293,38 @@ public void lock() throws SmartcarException {
     this.call("security", "POST", body);
   }
 
+
+  /**
+   * Send request to the /engine/oil endpoint
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/Vehicle.java]: Update this JavaDoc. It seems to be copy/pasted from the oil endpoint.

File: src/main/java/com/smartcar/sdk/data/BatchDeserializer.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package com.smartcar.sdk.data;
+
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import java.lang.reflect.Type;
+
+public class BatchDeserializer implements JsonDeserializer<BatchResponse> {
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/data/BatchDeserializer.java]: Put a JavaDoc on what this does

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.GsonBuilder;
+import org.apache.commons.text.CaseUtils;
+import java.util.HashMap;
+import java.util.Iterator;
+
+public class BatchResponse extends ApiData {
+  private HashMap<String, JsonObject> responseData = new HashMap<>();
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/data/BatchResponse.java]: Use the interface here:
```suggestion
  private Map<String, JsonObject> responseData = new HashMap<>();
```

You'll also need to import `java.util.Map`

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+  public BatchResponse(JsonArray responses) {
+
+    Iterator<JsonElement> responseIterator = responses.iterator();
+
+    while (responseIterator.hasNext()){
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/data/BatchResponse.java]: Nit consistency on spacing.
```suggestion
    while (responseIterator.hasNext()) {
```

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java (hunk 1 above)
Comment (sankethkatta) [reply src/main/java/com/smartcar/sdk/data/BatchResponse.java]: I think we could wrap everything into the `SmartcarException` type.

File: src/test/java/com/smartcar/sdk/VehicleTest.java [hunk 12]
```diff
@@ -259,4 +264,52 @@ public void testLock() throws Exception {
 
     this.subject.lock();
   }
+
+  @Test
+  public void testBatch() throws Exception {
```
Comment (sankethkatta) [inline src/test/java/com/smartcar/sdk/VehicleTest.java]: we should add tests for some failure cases as well.

also add cases to test for the imperial header being set

File: src/main/java/com/smartcar/sdk/data/BatchDeserializer.java (hunk 9 above)
Comment (theodorewahle) [reply src/main/java/com/smartcar/sdk/data/BatchDeserializer.java]: added

File: src/main/java/com/smartcar/sdk/Vehicle.java (hunk 8 above)
Comment (theodorewahle) [reply src/main/java/com/smartcar/sdk/Vehicle.java]: oopsie

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java (hunk 1 above)
Comment (theodorewahle) [reply src/main/java/com/smartcar/sdk/data/BatchResponse.java]: done!

File: src/test/java/com/smartcar/sdk/VehicleTest.java (hunk 12 above)
Comment (theodorewahle) [reply src/test/java/com/smartcar/sdk/VehicleTest.java]: added imperial test

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java [hunk 13]
```diff
@@ -38,13 +40,14 @@ private static String toCamelCase(String fieldName) {
...
     JsonObject res = this.responseData.get(path);
-
     Integer code = res.get("code").getAsInt();
+    JsonObject body = res.get("body").getAsJsonObject();
+
     if (code != 200) {
-      // TODO : handle error
-      System.out.println("ERROR");
+      String errorMessage = body.get("message").getAsString();
+      throw new SmartcarException(errorMessage);
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/data/BatchResponse.java]: We should use the longer constructor or our Factory to parse the errorCode/statusCode, ...
```java
public SmartcarException(int statusCode, String error, String message, String code)
```
or 
```java
  public static SmartcarException Factory(final Response response) throws IOException {
```

Take a look here: https://github.com/smartcar/java-sdk/blob/master/src/main/java/com/smartcar/sdk/SmartcarException.java#L22

File: src/main/java/com/smartcar/sdk/Vehicle.java [hunk 14]
```diff
@@ -295,13 +291,14 @@ public void lock() throws SmartcarException {
 
 
   /**
-   * Send request to the /engine/oil endpoint
+   * Send request to the /batch endpoint
    *
-   * @return the engine oil status of the vehicle
+   * @return the BatchResponse object containing the response
```
Comment (sankethkatta) [inline src/main/java/com/smartcar/sdk/Vehicle.java]: also add the parameter description above the returns:

```
@param paths the paths of endpoints to send requests to (ex. "/odometer", "/location", ...)
```

File: src/main/java/com/smartcar/sdk/data/BatchResponse.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+
+  private static String toCamelCase(String fieldName) {
+    if (fieldName.contains("_")) { // checks for snake case
+      return CaseUtils.toCamelCase(fieldName, false, new char[]{'_'});
+    }
+    return fieldName;
+  }
+
+  static GsonBuilder gson = new GsonBuilder().setFieldNamingStrategy((field) -> toCamelCase(field.getName()));
```
Comment (adolfoportilla) [inline src/main/java/com/smartcar/sdk/data/BatchResponse.java]: can you move this to the top of the class?

File: src/test/java/com/smartcar/sdk/VehicleTest.java [hunk 16]
```diff
@@ -259,4 +282,102 @@ public void testLock() throws Exception {
...
+    BatchResponse expectedBatch = new BatchResponse(parsed.getAsJsonArray());
+    SmartcarResponse<BatchResponse> res = new SmartcarResponse(expectedBatch);
+    ArrayList<String> paths = new ArrayList<>();
+    paths.add("/odometer");
+    PowerMockito.doReturn(res).when(this.subject, "call", eq("batch"), eq("POST"), refEq(body), refEq(BatchResponse.class));
+
+    BatchResponse batch = this.subject.batch(paths);
+
+    SmartcarResponse<VehicleOdometer> odo = batch.get("/odometer");
+    VehicleOdometer expectedOdo = new VehicleOdometer(32768);
```
Comment (adolfoportilla) [inline src/test/java/com/smartcar/sdk/VehicleTest.java]: is there an easy way to add this number (32768), to the expected response that way we don't repeat the value twice?

File: src/test/java/com/smartcar/sdk/VehicleTest.java [hunk 17]
```diff
@@ -259,4 +282,102 @@ public void testLock() throws Exception {
...
+          )
+          .add(Json.createObjectBuilder()
+            .add("path", "/fuel")
+            .build()
+          )
+          .build()
+      )
+      .build();
+    RequestBody body = RequestBody.create(Vehicle.JSON, json.toString());
+    String expectedJson = "[{\"headers\":{\"unitSystem\":\"metric\"},\"path\":\"/odometer\",\"code\":200,\"body\":{\"distance\":32768}}]";
```
Comment (adolfoportilla) [inline src/test/java/com/smartcar/sdk/VehicleTest.java]: is there another way to write the expected JSON that it's easier to maintain? (i.e. using a builder to construct this response).

File: gradle.properties [hunk 18]
```diff
@@ -1,6 +1,6 @@
 libGroup=com.smartcar.sdk
 libName=java-sdk
-libVersion=2.1.6
+libVersion=2.1.7
```
Comment (theodorewahle) [inline gradle.properties]:
```suggestion
libVersion=2.2.0
```

File: src/test/java/com/smartcar/sdk/VehicleTest.java (hunk 16 above)
Comment (theodorewahle) [reply src/test/java/com/smartcar/sdk/VehicleTest.java]: I'm not sure what you mean. My understanding was that this _is_ the expected response.

File: src/test/java/com/smartcar/sdk/VehicleTest.java (hunk 17 above)
Comment (theodorewahle) [reply src/test/java/com/smartcar/sdk/VehicleTest.java]: yes, moved to seperate file!

File: src/test/java/com/smartcar/sdk/VehicleTest.java (hunk 16 above)
Comment (theodorewahle) [reply src/test/java/com/smartcar/sdk/VehicleTest.java]: This is now sort of a moot point now that the string is in separate file

**Pull request merged**